optimum/habana/transformers/gaudi_configuration.py

```diff
@@ -57,6 +57,10 @@ def __init__(self, **kwargs):
         self.hmp_bf16_ops = kwargs.pop("hmp_bf16_ops", DEFAULT_BF16_OPS)
         self.hmp_fp32_ops = kwargs.pop("hmp_fp32_ops", DEFAULT_FP32_OPS)
         self.hmp_is_verbose = kwargs.pop("hmp_is_verbose", False)
+        self.use_torch_autocast = kwargs.pop("use_torch_autocast", False)
+
+        if self.use_habana_mixed_precision and self.use_torch_autocast:
+            raise ValueError("`use_habana_mixed_precision` and `use_torch_autocast` cannot be both `True` in your Gaudi configuration, you must choose one or the other to perform mixed-precision training.")
```

Comment:
Do I understand correctly that anything related to autocast should be removed from GaudiConfig? For now not every model is fit to run under autocast context. I think the same stands for hmp. Not every model uses it. Should we somehow note that specific model shouldn't be used with autocast or maybe it would be good to leave it up to GaudiConfig?

Reply:
Yes you're right, I went a bit fast. Leaving it up to the Gaudi config sounds good to me, and that way users shouldn't have to care about this unless they want to go at a lower level.
So let's keep `use_torch_autocast` in the Gaudi configuration. And then, at trainer initialization, if `use_torch_autocast` is True we can set `args.bf16 = True` and also raise the error there if `use_habana_mixed_precision` is True too. WDYT?

Reply:
I don't think we can enable bf16 argument without deleting it like [here](https://github.dev/jwieczorekhabana/optimum-habana/blob/9fdd92ec6f4eba0b4d98fd752ed207a067426340/optimum/habana/transformers/trainer.py#L149) , because it would be propagated to parent HF Trainer. Correct me if I'm wrong but I think that it would cause error [here](https://github.com/huggingface/transformers/blob/a7920065f2cfd2549b838f9a30afd7c265fcdd88/src/transformers/trainer.py#LL611C16-L611C16). And were we to not delete args.half_precision_backend I think it would throw an error [here](https://github.com/huggingface/transformers/blob/a7920065f2cfd2549b838f9a30afd7c265fcdd88/src/transformers/trainer.py#L652).

I think it's better to familiarize users with autocast usage on Gaudi by setting half_precision_backend=hpu_amp (as we can't utilize auto option) rather than adding another argument that needs to be used and works only as a dummy.

Reply:
Well the `--bf16` argument is the one that actually enables mixed-precision training in Transformers. `half_precision_backend` is just there to specify the backend and doesn't enable anything.

I'm okay with not using it. But in that case, users are not really going to use `half_precision_backend` either since it will default to `hpu_amp` (unless they want to use the CPU, which should almost never happen). They will just control it with the `use_torch_autocast` field of the Gaudi config.

Reply:
Regarding the errors that would be triggered with `bf16` in the parent init, I think we should just set `args.bf16 = False` right before calling `super().__init__` and set it back to its value right after. I know it's very dirty, but 1) we don't need the code in the `if args.bf16` blocks to be run, 2) since we constantly adapt to Transformers & co, things are not always pretty and I prefer having something easier to maintain even though it's dirty.